x/sync/syncable_db_server.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package sync
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+
+	syncpb "github.com/ava-labs/avalanchego/proto/pb/sync"
+)
+
+var _ syncpb.SyncableDBServer = (*SyncableDBServer)(nil)
+
+func NewSyncableDBServer(db SyncableDB) *SyncableDBServer {
+	return &SyncableDBServer{db: db}
+}
+
+type SyncableDBServer struct {
+	syncpb.UnsafeSyncableDBServer
+
+	db SyncableDB
+}
+
+func (s *SyncableDBServer) GetMerkleRoot(
+	ctx context.Context,
+	_ *emptypb.Empty,
+) (*syncpb.GetMerkleRootResponse, error) {
+	root, err := s.db.GetMerkleRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &syncpb.GetMerkleRootResponse{
+		RootHash: root[:],
+	}, nil
+}
+
+func (s *SyncableDBServer) GetChangeProof(
+	ctx context.Context,
+	req *syncpb.GetChangeProofRequest,
+) (*syncpb.ChangeProof, error) {
+	startRootID, err := ids.ToID(req.StartRootHash)
+	if err != nil {
+		return nil, err
+	}
+	endRootID, err := ids.ToID(req.EndRootHash)
+	if err != nil {
+		return nil, err
+	}
+	changeProof, err := s.db.GetChangeProof(
+		ctx,
+		startRootID,
+		endRootID,
+		req.StartKey,
+		req.EndKey,
+		int(req.KeyLimit),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return changeProof.ToProto(), nil
+}
+
+func (s *SyncableDBServer) VerifyChangeProof(
+	ctx context.Context,
+	req *syncpb.VerifyChangeProofRequest,
+) (*syncpb.VerifyChangeProofResponse, error) {
+	var proof merkledb.ChangeProof
+	if err := proof.UnmarshalProto(req.Proof); err != nil {
+		return nil, err
+	}
+
+	rootID, err := ids.ToID(req.ExpectedRootHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO there's probably a better way to do this.
```

Comment:
The issue is that there are a lot of potential errors that `s.db.VerifyChangeProof` can return. We could enumerate all of them and use a mapping, but that creates a dependency on the database implementation, and if we change the database implementation such that it returns a new kind of error, we'd have to update the mapping/error list. Or am I missing something and this is still a doable thing?

I looked at `database/rpcdb` and originally wanted to do a similar thing, but for that, the mapping is super easy: if it isn't `database.ErrClosed` or `database.ErrNotFound`, it's unexpected and the server method should error. In this case, it's not so simple, as described above.

Reply:
right its worth consideration I dont think every error requires a custom error type but in the case where a specific error has special meaning you can define it with an appropriate code. Otherwise it defaults to `Other`. I don't have a deep understanding of scope of errors here but it would not require a 1:1 mapping. Regardless not blocking this PR.